mne/utils/numerics.py

```diff
@@ -734,22 +736,28 @@ def object_diff(a, b, pre=''):
             if key not in k2s:
                 out += pre + ' right missing key %s\n' % key
             else:
-                out += object_diff(a[key], b[key], pre + '[%s]' % repr(key))
+                out += object_diff(a[key], b[key],
+                                   pre=(pre + '[%s]' % repr(key)),
+                                   ignore_nan=ignore_nan)
     elif isinstance(a, (list, tuple)):
         if len(a) != len(b):
             out += pre + ' length mismatch (%s, %s)\n' % (len(a), len(b))
         else:
             for ii, (xx1, xx2) in enumerate(zip(a, b)):
-                out += object_diff(xx1, xx2, pre + '[%s]' % ii)
+                out += object_diff(xx1, xx2, pre + '[%s]' % ii, ignore_nan)
     elif isinstance(a, (str, int, float, bytes, np.generic)):
         if a != b:
             out += pre + ' value mismatch (%s, %s)\n' % (a, b)
     elif a is None:
         if b is not None:
             out += pre + ' left is None, right is not (%s)\n' % (b)
     elif isinstance(a, np.ndarray):
-        if not np.array_equal(a, b):
-            out += pre + ' array mismatch\n'
+        if ignore_nan:
+            if not np.array_equal(a[~np.isnan(a)], b[~np.isnan(b)]):
+                out += pre + ' array mismatch\n'
+        else:
+            if not np.array_equal(a, b):
+                out += pre + ' array mismatch\n'
```

Comment:
```
a_idx = ~np.isnan(a) if ignore_nan else slice(None)
b_idx = ~np.isnan(b) if ignore_nan else slice(None)
if not np.array_equal(a[a_idx], b[b_idx]):
    out += ...
```

Reply:
And actually the code you wrote will say that these are the same:
```
a = [1, np.nan, 0]
b = [np.nan, 1, 0]
```
Which it will not.

I would use the `try/except` pattern and rely on NumPy to do the right thing:

https://stackoverflow.com/questions/10710328/comparing-numpy-arrays-containing-nan#answer-10710613

Reply:
we can always compare `a_idx` and `b_idx` aswell